File: app/src/main/java/com/shizq/bika/utils/TimeUtil.java
package com.shizq.bika.utils;

import android.annotation.SuppressLint;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.TimeZone;

@SuppressLint("SimpleDateFormat")
public class TimeUtil {
    //反编译源码 从里面复制的
    public static String B(String str) {
        if (str == null||str.equals("")) {
            return "";
        }
        Calendar instance = Calendar.getInstance();
        try {
            SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
            simpleDateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
            instance.setTime(simpleDateFormat.parse(str));
            Calendar instance2 = Calendar.getInstance();
            instance2.add(13, -60);
            if (instance2.compareTo(instance) <= 0) {
                instance2.add(13, 60);
                return ((instance2.getTimeInMillis() - instance.getTimeInMillis()) / 1000) + "秒前";
            }
            Calendar instance3 = Calendar.getInstance();
            instance3.add(12, -60);
            if (instance3.compareTo(instance) <= 0) {
                instance3.add(12, 60);
                return ((instance3.getTimeInMillis() - instance.getTimeInMillis()) / 60000) + "分钟前";
            }
            Calendar instance4 = Calendar.getInstance();
            if (instance4.get(1) == instance.get(1) && instance4.get(6) == instance.get(6)) {
                SimpleDateFormat simpleDateFormat2 = new SimpleDateFormat("HH:mm");
                return "今天 " + simpleDateFormat2.format(instance.getTime());
            }
            Calendar instance5 = Calendar.getInstance();
            instance5.add(6, -1);
            if (instance5.get(1) == instance.get(1) && instance5.get(6) == instance.get(6)) {
                SimpleDateFormat simpleDateFormat3 = new SimpleDateFormat("HH:mm");
                return "昨天 " + simpleDateFormat3.format(instance.getTime());
            } else if (Calendar.getInstance().get(1) == instance.get(1)) {
                SimpleDateFormat simpleDateFormat4 = new SimpleDateFormat("M月d日 HH:mm");
                return "" + simpleDateFormat4.format(instance.getTime());
            } else {
                SimpleDateFormat simpleDateFormat5 = new SimpleDateFormat("yyyy年MM月dd日 HH:mm");
                return "" + simpleDateFormat5.format(instance.getTime());
            }
        } catch ( ParseException e) {
            e.printStackTrace();
            return "";
        }
    }

}

File: app/src/main/java/com/shizq/bika/utils/TimeUtil.kt
package com.shizq.bika.utils

import android.annotation.SuppressLint
import java.text.SimpleDateFormat
import java.util.*


@SuppressLint("SimpleDateFormat")
class TimeUtil {

    private fun getDate(date: Date): String {
        //传入的时间
        val calendar = Calendar.getInstance()
        calendar.time = date

        //秒
        val second = Calendar.getInstance()
        second.add(Calendar.SECOND, -60)
        if (second <= calendar) {
            second.add(Calendar.SECOND, 60)
            return ((second.timeInMillis - calendar.timeInMillis) / 1000).toString() + "秒前"
        }

        //分
        val minute = Calendar.getInstance()
        minute.add(Calendar.MINUTE, -60)
        if (minute <= calendar) {
            minute.add(Calendar.MINUTE, 60)
            return ((minute.timeInMillis - calendar.timeInMillis) / 60000).toString() + "分钟前"
        }

        //今天
        val day = Calendar.getInstance()
        if (day[Calendar.YEAR] == calendar[Calendar.YEAR]
            && day[Calendar.DAY_OF_YEAR] == calendar[Calendar.DAY_OF_YEAR]
        ) {
            val simpleDateFormat2 = SimpleDateFormat("HH:mm")
            return "今天 " + simpleDateFormat2.format(calendar.time)
        }

        //其他时间
        val yesterday = Calendar.getInstance()
        yesterday.add(Calendar.DAY_OF_YEAR, -1)
        return if (yesterday[Calendar.YEAR] == calendar[Calendar.YEAR]
            && yesterday[Calendar.DAY_OF_YEAR] == calendar[Calendar.DAY_OF_YEAR]
        ) {
            "昨天 " + SimpleDateFormat("HH:mm").format(calendar.time)
        } else if (Calendar.getInstance()[Calendar.YEAR] == calendar[Calendar.YEAR]) {
            SimpleDateFormat("M月d日 HH:mm").format(calendar.time)
        } else {
            SimpleDateFormat("yyyy年MM月dd日 HH:mm").format(calendar.time)
        }
    }

    fun time(stringDate: String): String {
        val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
        simpleDateFormat.timeZone = TimeZone.getTimeZone("UTC")
        return getDate(simpleDateFormat.parse(stringDate)!!)
    }

    fun getDate(milSecond: Long): String {
        val date = Date(milSecond)
        return getDate(date)
    }

    //2000-01-01T00:00:00+00:00 转换成 2000-01-01 00:00:00
    fun getDate(StringDate: String): String {
        val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX")
        val date = simpleDateFormat.parse(StringDate)
        val simpleDateFormat2 = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        return simpleDateFormat2.format(date!!)
    }

    //2000-01-01T00:00:00.000Z 转换成 2000-01-01
    fun getBirthday(StringDate: String): String {
        val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
        val date = simpleDateFormat.parse(StringDate)
        val simpleDateFormat2 = SimpleDateFormat("yyyy-MM-dd")
        return simpleDateFormat2.format(date!!)
    }

    fun registerDays(day: Int): Boolean {
        val createdAt = SPUtil.get("user_created_at", "") as String
        val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
        val time = simpleDateFormat.parse(createdAt)!!.time
        val currentTime = System.currentTimeMillis()
        return (currentTime - time) / (1000 * 60 * 60 * 24) >= day
    }
}